### PR TITLE
🚧Create text box control

### LINKED ISSUE #78: 🚧Create text box control
### I have done the items below . . .

- [X] I have updated the title by replacing the '**_<title_**>' section.

### Description

Create a text box control that gives the user the ability to render text in a rectangular space.

This control will be complicated to make.  The idea here is to utilize the `Label` control for the text.  The control will need to detect keyboard input so it can change the text of the label control to be displayed to the user.  Deletion capability will also be required using the _**Delete**_ and _**Backspace**_ keys.

For text selection, a solid rectangle should be rendered behind the text. The color of the text should be set to the inverse color of the selection rectangle.  The inverse color will need to be calculated somehow.

### Acceptance Criteria

```[tasklist]
### Tasks
- [x] Text rendered in a rectangular area
- [x] Hitting _**Delete**_ key will delete characters that are to the right of the blinking cursor
- [x] Hitting _**Backspace**_ key will delete characters that are to the left of the blinking cursor
- [x] Hitting the _**Home**_ key by itself will move the cursor to the beginning
- [x] Hitting the _**End**_ key by itself will move the cursor to the end
- [x] Ability to set the text color
- [x] Ability to set the color of the selection
- [x] Ability to set the color of the cursor
- [x] Add a new method overload for the `IFont.GetCharacterBounds()` method to take a `StringBuilder` param
```

```[tasklist]
### Text Selection
- [x] Holding shift while using the **_Right Arrow_** key will select text one character at a time to the right of where the cursor started during the selection process
- [x] Holding shift while using the **_Left Arrow_** key will select text one character at a time to the left of where the cursor started during the selection process
- [x] Holding shift while using the **_End_** key will select all of the text to the right of where the cursor started during the selection process
- [x] Holding shift while using the **_Home_** key will select all of the text to the left of where the cursor started during the selection process
- [x] When text is selected, a filled rectangle will be rendered with the text on top.  This shows the text as selected.
```

### ToDo Items

```[tasklist]
### Tasks
- [x] Draft pull request created and linked to this issue
- [x] Issue linked to the proper project
- [x] Priority label added to issue (**_low priority_**, **_medium priority_**, or **_high priority_**)
- [x] Issue linked to proper milestone
- [x] Unit tests have been written and/or adjusted for code additions or changes
- [x] All unit tests pass
```

### Issue Dependencies

- #75
- #76

### Related Work

_No response_

### PR DESCRIPTION
**Pull request by CalvinWilkinson**
<h1 style="font-weight:bold" align="center">Preview Feature Pull Request</h1>
<h2 style="font-weight:bold" align="center">📋Review Checklist📋</h2>

- [x] The **_head(source)_** branch for this pull request is a **_preview feature_** branch, with the correct naming syntax, in the following example:
  - A **_head(source)_** branch is the branch being merged into a target branch.
    <details closed><summary>Example</summary>

      ``` xml
      Syntax: preview/feature/<issue-num>-<description>
      Example: preview/feature/123-my-preview-feature
      ```
    </details>

- [x] The **_head(source)_** branch for this pull request is created from a **_preview release_** branch, with the correct naming syntax, in the following example:
  <details closed><summary>Example</summary>

    ``` xml
    Syntax: preview/v<major>.<minor>.<patch>-preview.<prev-num>
    Example: preview/v1.2.3-preview.4
    ```
  </details>

- [x] The **_base(target)_** branch for this pull request is a **_preview release_** branch, with the correct naming syntax, in the following example:
  - A **_base(target)_** branch is the branch that the **_head(source)_** branch is merging into.

    <details closed><summary>Example</summary>

      ``` xml
      Syntax: preview/v<major>.<minor>.<patch>-preview.<prev-num>
      Example: preview/v1.2.3-preview.4
      ```
    </details>

💡For more information on branching, refer to the project's [branching documentation](https://github.com/KinsonDigital/.github/blob/master/docs/BranchingFullGitFlow.md).

- [x] Associated issue exists and is linked to this pull request.
  - One issue per pull request.

- [x] Pull request title matches the title of the associated issue.

- [x] A `preview` label has been added to the pull request.

- [x] The labels attached to this PR match the labels attached to the associated issue.

- [x] Project assigned matches the project of the associated issue.

- [x] Milestone assigned matches the milestone of the associated issue.

- [x] My code follows the coding style of this project.
  - The style is enforced by the **_.editorconfig_** files in the project and displayed as warnings.  If there is an edge case, reach out, and let's discuss it.

- [x] I have manually tested my code changes to the best of my ability.

- [x] All tests passed locally.
  - This is required because unit tests might pass locally but not in the CI environment during the status check process or vice-versa.
  - Tests might pass on the developer's machine but not necessarily on the code reviewer's machine.
  - Status checks run for this pull request when merging into the **_preview/v\*.\*.\*-preview.\*_** branch.  These status checks run every time a change is pushed to the pull request.  These checks validate version syntax, tagging, builds, unit tests, and more.

### CODE REVIEW

**Comment by CalvinWilkinson** [comment]
> **Note** Added https://github.com/KinsonDigital/Velaptor/labels/%F0%9F%A7%A8breaking%20changes
> - Renamed the `MousePositionEventArgs` struct to `MouseMoveEventArgs`.
> - Renamed the `MouseMoveEventArgs`.MousePosition` property to `MouseMoveEventArgs.LocalPos`
> - Changed the `MouseMoveEventArgs` from a class to a `readonly record struct`

**Comment by CalvinWilkinson** [comment]
> **Note** Add https://github.com/KinsonDigital/Velaptor/labels/%E2%9C%A8new%20feature 's
> Added a property named `GlobalPos` to the `MouseMoveEventArgs` struct to hold the global position relative to the top left corner of the window.

**Comment by CalvinWilkinson** [comment]
> [!Note]
> Fixed the following bugs
> - Fixed a bug with the `Font` class where the bounds width for the space character was not being calculated correctly.  These bounds can be returned from the `Font.GetCharacterBounds()` method.
> - Fixed a bug where measuring text with `Font.Measure()` was not measuring spaces at the end of the text.
> - Fixed issues with UI control `MouseDown` and `MouseUp` events not firing
> - Fixed a small issue where if text ended with only spaces, they would not be rendered
> - Fixed a bug where the internal text size cache of the `Font` class was not clearing the cache when changing font sizes.
> - Fixed a bug where Velaptor would crash when trying to measure the same piece of text content more than once.

File: .vscode/settings.json
Comment (KinsonDigitalAdmin) [inline .vscode/settings.json:1]: Add empty line to end of the file

File: Velaptor/Graphics/Renderers/FontRenderer.cs [hunk 1]
```diff
@@ -147,36 +214,77 @@ protected override void ShutDown()
...
+            metricsWithClr.ToArray(),
+            font,
+            origin,
+            renderSize,
+            angle,
+            atlasWidth,
+            atlasHeight);
+    }
+
+     /// <summary>
```
Comment (KinsonDigitalAdmin) [inline Velaptor/Graphics/Renderers/FontRenderer.cs:244]: Align code docs

**Review by KinsonDigitalAdmin — CHANGES_REQUESTED** [review]

**Review by KinsonDigitalAdmin — DISMISSED** [review]

File: Velaptor/UI/ControlBase.cs [hunk 2]
```diff
@@ -142,18 +194,115 @@ public virtual void Update(FrameTime frameTime)
...
+        foreach (var downKey in currState.GetDownKeys())
+        {
+            OnKeyDown(downKey);
+        }
+    }
+
+    /// <summary>
+    /// Process any keyboard key up events.
+    /// </summary>
+    /// <param name="currState">The current state of the keyboard this frame.</param>
```
Comment (kselena) [inline Velaptor/UI/ControlBase.cs:274]: Reword - add "in"
```suggestion
    /// <param name="currState">The current state of the keyboard in this frame.</param>
```

File: Velaptor/UI/ControlBase.cs [hunk 3]
```diff
@@ -142,18 +194,115 @@ public virtual void Update(FrameTime frameTime)
...
+    /// <param name="prevState">The previous state of the keyboard the previous frame.</param>
```
Comment (kselena) [inline Velaptor/UI/ControlBase.cs:275]: Reword - add "in"
```suggestion
    /// <param name="prevState">The previous state of the keyboard in the previous frame.</param>
```

File: Velaptor/UI/TextBox.cs [hunk 4]
```diff
@@ -1,78 +1,371 @@
...
     /// Loads the content of the <see cref="TextBox"/>.
     /// </summary>
-    public override void LoadContent() => this.text.LoadContent();
+    public override void LoadContent()
+    {
+        var contentLoader = ContentLoaderFactory.CreateContentLoader();
+
+        // TODO: Need to add ability to change font to the text box.  This means being able to reload
+        // the font when the font type or size changes.
+        this.font = contentLoader.LoadFont("TimesNewRoman-Regular", 12);
```
Comment (kselena) [inline Velaptor/UI/TextBox.cs:341]: Do you still need this TODO?
```suggestion
        // TODO: Need to add ability to change font to the text box.  This means being able to reload
        // the font when the font type or size changes.
        this.font = contentLoader.LoadFont("TimesNewRoman-Regular", 12);
```

File: Velaptor/UI/TextBox.cs [hunk 5]
```diff
@@ -106,23 +401,762 @@ public override void Render()
...
 
-        // this.text.Render(renderer);
+        this.shapeRenderer.Render(this.ctrlBorder, layer: 0);
+
+        if (this.textCursor.Visible)
+        {
+            this.shapeRenderer.Render(this.textCursor.Cursor, layer: 0);
+        }
+
+        // and we are unselecting text to the left, we cannot see any of the selection rect.
```
Comment (kselena) [inline Velaptor/UI/TextBox.cs:411]: Reword? 
```suggestion
        // When we are unselecting text to the left, we cannot see any of the selection rect.
```

File: Velaptor/UI/TextBox.cs [hunk 6]
```diff
@@ -106,23 +401,762 @@ public override void Render()
...
+        UpdateBounds(charKey); // NEW
+        CalcYPositions(prevHeight, this.textSize.Height);
+
+        this.currentCharIndex = charIndexAtEnd
+            ? this.text.LastCharIndex()
+            : this.currentCharIndex + 1;
+    }
+
+    /// <summary>
+    /// Removes the a character from the text box.
```
Comment (kselena) [inline Velaptor/UI/TextBox.cs:673]: Grammar - If you're talking about a specific char, it's best to set it apart so the sentence flows.  You can also use "a character" if that's the actual name. 
```suggestion
    /// Removes the 'a' character from the text box.
```

**Review by kselena — CHANGES_REQUESTED** [review]

File: Testing/VelaptorTests/InternalExtensionMethodsTests.cs [hunk 7]
```diff
@@ -1640,6 +2530,105 @@ public void ToBatchItem_WithCircleShapeOverload_ReturnsCorrectResult()
...
+        // Assert
+        actual.Should().Be(expected);
+    }
+
+    [Theory]
+    [MemberData(nameof(IsCtrlKeyTestData))]
+    public void IsArrowKeIsCtrlKey_WhenInvoked_ReturnsCorrectResult(KeyCode key, bool expected)
+    {
+        // Arrange & Act
+        var actual = key.IsCtrlKey();
```
[the remark this answers is not in the document]
Comment (KinsonDigitalAdmin) [reply Testing/VelaptorTests/InternalExtensionMethodsTests.cs:2605]: The single line `var actual = key.IsCtrlKey()` is the arrangement AND the act all in one line. This is why it says `// Arrange & Act`. The key part is the `&` part.

Good catch on the name of the unit test method though!!

File: Velaptor/UI/MouseMoveEventArgs.cs [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+    /// <summary>
+    /// Gets the position of the mouse relative to the top left corner of the window.
+    /// </summary>
+    public Point GlobalPos { get; }
+
+    /// <summary>
+    /// Gets the position of the mouse relative to the top left corner of the control.
+    /// </summary>
+    public Point LocalPos { get; }
```
[the remark this answers is not in the document]
Comment (KinsonDigitalAdmin) [reply Velaptor/UI/MouseMoveEventArgs.cs:38]: It is fine the way it is.  It is what it says that it is.

The global position represents the position of the mouse in the window.  Controls live in this window space. So when you talk about a control position, it is the position of the control in its own "world" which is the window.  

Local position is the position of the mouse _**RELATIVE**_ to the top left corner of the control itself, not the window.

image

In this image below, the distance in orange is the horizontal distance relative to the left side of the window.  If the mouse was touching the left side of the button, this would be the _**global**_ mouse position.  

The yellow is the distance from the left side of the button to the mouse cursor.  This would be the _**local**_ mouse position.

**Review by kselena — DISMISSED** [review]

**Review by KinsonDigitalAdmin — DISMISSED** [review]

File: Testing/VelaptorTests/ExtensionMethods/TextBoxExtensionsTests.cs [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+namespace VelaptorTests.ExtensionMethods;
+
+using Velaptor.UI;
+
+/// <summary>
+/// Tests all of the extension methods related to the <see cref="TextBox"/> control.
+/// </summary>
+public class TextBoxExtensionsTests
+{
+    // TODO: Create tests
```
Comment (kselena) [inline Testing/VelaptorTests/ExtensionMethods/TextBoxExtensionsTests.cs:14]: Is this TODO still valid? 
```suggestion
    // TODO: Create tests
```

File: Velaptor/ExtensionMethods/GPUDataTypeExtensions.cs [hunk 10]
```diff
@@ -0,0 +1,564 @@
...
+
+        return gpuData;
+    }
+
+    /// <summary>
+    /// Updates the <see cref="RectVertexData.Color"/> of a vertex using the given <paramref name="vertexNumber"/>
+    /// for the given <paramref name="gpuData"/>.
+    /// </summary>
+    /// <param name="gpuData">The GPU data to update.</param>
+    /// <param name="color">The color to set the vertex to.</param>
```
Comment (kselena) [inline Velaptor/ExtensionMethods/GPUDataTypeExtensions.cs:497]: Grammar
```suggestion
    /// <param name="color">The color to set the vertex.</param>
```

File: Velaptor/ExtensionMethods/TextBoxExtensions.cs [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+
+        var left = charBounds.Min(cb => cb.bounds.Left);
+        var right = charBounds.Max(cb => cb.bounds.Right);
+        var width = Math.Abs(left - right);
+
+        return left + width.Half();
+    }
+
+    /// <summary>
+    /// Returns a value indicating if their is a gap between the right end of the text in the given <paramref name="charBounds"/>
```
Comment (kselena) [inline Velaptor/ExtensionMethods/TextBoxExtensions.cs:171]: Reword/spelling
```suggestion
    /// Returns a value indicating whether or not there is a gap between the right end of the text in the given <paramref name="charBounds"/>
```

**Review by kselena — CHANGES_REQUESTED** [review]

**Review by KinsonDigitalAdmin — APPROVED** [review]

**Review by kselena — DISMISSED** [review]

**Review by KinsonDigitalAdmin — DISMISSED** [review]

**Review by KinsonDigitalAdmin — APPROVED** [review]
